zetasql/fuzzing/component/arguments/argument.h

```diff
@@ -17,30 +17,35 @@
 #ifndef ZETASQL_FUZZING_ARGUMENT_H
 #define ZETASQL_FUZZING_ARGUMENT_H
 
+#include <string>
 #include <memory>
-
-#include "zetasql/fuzzing/component/function.h"
+#include "zetasql/fuzzing/component/fuzz_targets/fuzz_target.h"
 
 namespace zetasql_fuzzer {
 
 class Argument {
  public:
-  virtual void Accept(zetasql_fuzzer::Function& function) const = 0;
+  virtual void Accept(zetasql_fuzzer::FuzzTarget& function) = 0;
 };
 
-template <typename T>
-class PositionalArg : public Argument {
+template <typename ArgType>
+class TypedArg : public Argument {
  public:
-  PositionalArg(std::shared_ptr<const T> value) : argument(value) {}
-  virtual void Accept(zetasql_fuzzer::Function& function) const override {
-    function.Visit(*this);
-  }
-  const std::shared_ptr<const T> GetArgument() const { return argument; }
-  const uint8_t GetPosition() const { return position; }
+  TypedArg(const ArgType& value) : argument(std::make_unique<ArgType>(value)) {}
+  TypedArg(ArgType&& value) : argument(std::make_unique<ArgType>(value)) {}
+  // Use referece or value?
+  std::unique_ptr<ArgType> ReleaseArg() { return std::move(argument); }
```

Comment:
I'm not sure if move semantics is best represented here. Same as in the ProtoExprExtractor#Release. How do we properly describe a class methods that is meant to be only used once? Or should we simply fall back to copy semantics.

Reply:
So I'd potentially return a StatusOr instead of just a unique_ptr, and then have a boolean that says if the call is "valid".  If the call isn't valid, then you'll want to return an error status.

Reply:
I think you misunderstood the intention here. The call should always be valid. I worry about the copy or move semantics. 

The design here is basically that, the argument value will be transferred first from Extractor to Argument, then from Argument to FuzzTarget. Since FuzzTarget will be the only one that's actually _using_ the argument value, we don't have to retain a copy at Extract or Argument. 

Reason for using pointer is that 
1) argument value is created by Extractor but used by FuzzTarget, so the life span of the value is not local, 
2) unique_ptr conveniently enables the "Optional" semantics at FuzzTarget side, i.e., the optional argument for FuzzTarget. Since only the pointers are passed around, FuzzTarget can check if an optional argument exists by checking whether the pointer is null or not 
3) unique_ptr has move semantics so we can optimize execution by eliminating argument copying.
4) (*) unique_ptr supports polymorphism. If the argument value has many subclass we don't have to create an Argument type for each subclass, if we have polymorphism in the Argument class. (This is not supported in the current implementation, but is possible if we create a constructor `TypedArg(std::unique<ArgType>& ptr): argument(ptr) {}`)